ebs-nvme-mapping.sh

```diff
@@ -1,10 +1,25 @@
 #!/bin/bash
+# To be used with the udev rule: /etc/udev/rules.d/999-aws-ebs-nvme.rules
 
-PATH="${PATH}:/usr/sbin"
+if [[ ! -x /usr/sbin/nvme ]]; then
+  echo "ERROR: NVME tools not installed." >> /dev/stderr
+  exit 1
+fi
 
-for blkdev in $( nvme list | awk '/^\/dev/ { print $1 }' ) ; do
-  mapping=$(nvme id-ctrl --raw-binary "${blkdev}" | cut -c3073-3104 | tr -s ' ' | sed 's/ $//g')
-  if [[ "${mapping}" == /dev/* ]]; then
-    ( test -b "${blkdev}" && test -L "${mapping}" ) || ln -s "${blkdev}" "${mapping}"
-  fi
-done
+if [[ ! -b ${1} ]]; then
+  echo "ERROR: cannot find block device ${1}" >> /dev/stderr
+  exit 1
+fi
+
+# capture 32 bytes at an offset of 3072 bytes from the raw-binary data
+# not all block devices are extracted with /dev/ prefix
+# use `xvd` prefix instead of `sd`
+# remove all trailing space
+nvme_link=$( \
+  /usr/sbin/nvme id-ctrl --raw-binary "${1}" | \
+  /usr/bin/cut -c3073-3104 | \
+  /bin/sed 's/^\/dev\///g'| \
+  /bin/sed 's/^sd/xvd/'| \
```

Comment:
The target device name depends on the instance type and AMI being used. One cannot assume (enforce) that all `sdX` devices should turn into `xvdX` devices.

Let the user pass in their desired device name and respect it all the way through to the end, unless the underlying utilities (e.g. `udev`) tell us otherwise.

https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/device_naming.html

Reply:
Updated accordingly.

Reply:
Updated accordingly.